### ISSUE TITLE
Provide a public API to create polymorphic containers

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#729]
We have public factory methods for creating NamedDomainObjectContainers, we should have something similar for polymorphic containers. 

The PolymorphicDomainObjectContainer is part of the public API and a powerful tool for building extensible DSLs. Unfortunately, there is no public way to create one. We should add `Project.polymorphicContainer(Class baseType)`

We merge the concepts of `NamedDomainObjectContainer` and `PolymorphicDomainObjectContainer`. Every container can be and should be polymorphic.

Also consider the case where one just wants to add something to the base/default type.

e.g., if your container is any type of `Widget` and by default there's a `DefaultWidget`, but someone could implement a `SpecializedWidget`:

```
widgets {
   foo {} // implicit type DefaultWidget
   bar(DefaultWidget) {} // explicit
   special(SpecializedWidget) {} // explicit
}
```

Something we should also think about is a way to register a factory for all containers, as is available in the software model. This allows others to reuse the types in locations beyond where the plugin author happens to register factories.

**Comment by lacasseio** [comment]
Rereading the description of this issue, I think what I mean was probably related to this line:
> Every container can be and should be polymorphic.

I don't know what would be the impact of changing `Project.container` to return polymorphic container instead of introducing another method for that purpose. It's just a though, feel free to disregard.

**Comment by mkobit** [comment]
Does `ObjectFactory` satisfy this requirement now?

**Comment by mkobit** [comment]
Doesn't look like it. I think I misunderstood the `ObjectFactory` documentation.

Maybe this should also support the `ExtensiblePolymorphicDomainObjectContainer` type? It isn't clear from a plugin author perspective when or how I should use these.

The original description also has references to using in the sotware model, which is no longer relevant and may provide a different way forward for this.

**Comment by mkobit** [comment]
Plugin implementations are currently relying on the internals of Gradle to be able to use this feature.

For example, the [`JetBrains/gradle-idea-ext-plugin`](https://github.com/JetBrains/gradle-idea-ext-plugin) is using both the internal `org.gradle.api.internal.DefaultPolymorphicDomainObjectContainer` and required `org.gradle.internal.reflect.Instantiator` for it to work. See the source code [here](https://github.com/JetBrains/gradle-idea-ext-plugin/blob/423e75ff40e209f963eb84d24f942bfe1f47faca/src/main/groovy/org/jetbrains/gradle/ext/RunConfigurations.groovy#L21-L37))

**Comment by adammurdoch** [comment]
Yes, I would just add some new methods to `ObjectFactory`. I wouldn't try to make `newInstance()` work, except perhaps to let the caller know that they should use one of the other methods.

**Comment by mkobit** [comment]
@lacasseio I'd definitely love to get to it, but probably won't have time at least in the next couple weeks.

**Comment by mkobit** [comment]
I recently took a quick look at possibly doing this. I didn't go much further after exploring what the internals look like. I'm not really sure what type of API is desired as well as to what the desired end-user experience is supposed to be, based on what I see internally.

- [`PolymorphicDomainObjectContainer`](https://docs.gradle.org/5.4.1/javadoc/org/gradle/api/PolymorphicDomainObjectContainer.html) - I only tried playing around with [`AuthenticationContainer`](https://docs.gradle.org/5.4.1/javadoc/org/gradle/api/artifacts/repositories/AuthenticationContainer.html, but it seems to have an explicit set of types it supports creating. For example:

    ```kotlin
    interface DumbAuth : Authentication
    
    repositories {
        maven {
            authentication {
                create("basic", BasicAuthentication::class)
                create("MyAuth", DumbAuth::class)
            }
        }
    }
    ```

* `org.gradle.api.internal.DefaultPolymorphicDomainObjectContainer` - has a `registerDefaultFactory` and `registerFactory`, so it seems like it is mostly just an internal `ExtensiblePolymorphicDomainObjectContainer` type implementation that seems basically the same as `ExtensiblePolymorphicDomainObjectContainer`, except it is an internal hidden type of `PolymorphicDomainObjectContainer`.

- [`ExtensiblePolymorphicDomainObjectContainer`](https://docs.gradle.org/5.4.1/javadoc/org/gradle/api/ExtensiblePolymorphicDomainObjectContainer.html) - seems to be the type supporting additional type registrations explicitly

My goal in using these is to represent a collection of various supported "things" that a user can create, and then the plugin would automatically create tasks, outputs, configurations, whatever, based on the type of object in the container. The `PolymorphicDomainObjectContainer` doesn't, to me, have a straightforward path for a plugin author to make use of. `ExtensiblePolymorphicDomainObjectContainer` seems more straightforward as the plugin author could register object factories for each supported type. 

After some light investigation, I don't feel like I have a clear understanding of what would be needed and accepted into Gradle for this issue.

**Comment by lacasseio** [comment]
Thanks @mkobit for sharing your input on this. We are working on getting https://github.com/gradle/gradle/pull/9393 merged in for 5.5. It won't solve your immediate use case, but there are several steps before we get to a public API for creating those type.

**Comment by eskatos** [comment]
This was added back in 6.1
https://docs.gradle.org/6.1/release-notes.html#new-factory-methods
https://docs.gradle.org/6.1/userguide/custom_gradle_types.html#extensiblepolymorphicdomainobjectcontainer
